src/builder.rs

```diff
@@ -254,93 +217,347 @@ impl<'a> RtpPacketBuilder<'a> {
     /// Write this packet into `buf` without any validity checks.  Returns the number of bytes
     /// written.
-    pub fn write_unchecked(self, buf: &mut Vec<u8>) -> usize {
-        let start_len = buf.len();
+    pub fn write_into_unchecked<'b: 'a>(&self, buf: &'b mut [u8]) -> usize {
+        let mut writer = RtpPacketWriterMutSlice::new(buf);
+        self.write_unchecked(&mut writer)
+    }
 
-        let mut byte = 0x80; // rtp version 2
-        if self.padding.is_some() {
-            byte |= 0x20;
+    /// Write this packet into `buf`.  On success returns the number of bytes written or an
+    /// `RtpWriteError` on failure.
+    pub fn write_into<'b: 'a>(&self, buf: &'b mut [u8]) -> Result<usize, RtpWriteError> {
+        let mut writer = RtpPacketWriterMutSlice::new(buf);
+        self.write(&mut writer)
+    }
+
+    /// Write this packet into `buf` without any validity checks.  The data will be appended to the
+    /// end of the provide Vec.
+    pub fn write_into_vec_unchecked(&self, buf: &'a mut Vec<u8>) {
+        let mut writer = RtpPacketWriterMutVec::new(buf);
+        self.write_unchecked(&mut writer)
+    }
+
+    /// Write this packet into `buf`.  The data will be appended to the end of the provide Vec.
+    pub fn write_into_vec(&self, buf: &'a mut Vec<u8>) -> Result<(), RtpWriteError> {
+        let mut writer = RtpPacketWriterMutVec::new(buf);
+        self.write(&mut writer)
+    }
+
+    /// Write this packet into a newly generated `Vec<u8>` without any validity checks.
+    pub fn write_vec_unchecked(&self) -> Vec<u8> {
+        let mut writer = RtpPacketWriterVec::default();
+        self.write_unchecked(&mut writer)
+    }
+
+    /// Write this packet into a newly generated `Vec<u8>`.
+    pub fn write_vec(&self) -> Result<Vec<u8>, RtpWriteError> {
+        let mut writer = RtpPacketWriterVec::default();
+        self.write(&mut writer)
+    }
+}
+
+/// Trait to provide the length of a piece of data in bytes
+pub trait PayloadLength {
+    fn len(&self) -> usize;
+    fn is_empty(&self) -> bool {
+        self.len() != 0
+    }
+}
+
+/// Trait to write an RTP packet into and/or from custom data types
+pub trait RtpPacketWriter {
+    type Output;
+    type Payload: PayloadLength;
+    type Extension: PayloadLength;
+
+    /// Reserve a number of bytes in the output.  Multiple calls are possible and provide the
+    /// entire size to reserve
+    fn reserve(&mut self, _size: usize) {}
+
+    /// Return the maximum size of the output.  If the output data is not bound to a fixed size,
+    /// `None` should be returned.
+    fn max_size(&self) -> Option<usize> {
+        None
+    }
+
+    /// Provides data to append to the output.  May be called multiple times per packet.
+    fn push(&mut self, data: &[u8]);
+
+    /// Provides the extension data to add to the output.  The extension should be written as-is
+    /// without any transformations.
+    fn push_extension(&mut self, extension_data: &Self::Extension);
```

Comment:
I think this is going to help me a lot for implementation extension support in the (de)payloader base classes.